File: ballattax/BALLATTAX.js
// BALLATTAX.js - Arcaqjs/QuickJS port of main.lua

import { switchScene, currentScene, inputs } from './scene_manager.js';
import menuScene from './menu.js';

// Lifecycle callbacks
export function enter(args = {}) {
    window.color(0x000000FF); // Set background color
    switchScene(menuScene, args);
}

export function input(evt, device, id, value, value2) {
    // update inputs state, evt describes a state change:
    while (device >= inputs.length) {
        inputs.push({ axes: [0, 0], buttons: [0, 0, 0, 0, 0, 0, 0, 0] });
    }
    if (evt === 'axis') {
        inputs[device].axes[id] = value;
    } else if (evt === 'button') {
        inputs[device].buttons[id] = value;
    }
    if (currentScene && currentScene.input) {
        currentScene.input(evt,device,id,value,value2);
    }
}

export function update(deltaT) {
    if (currentScene && currentScene.update) {
        return currentScene.update(deltaT, inputs);
    }
    return false;
}

export function draw(gfx) {
    if (currentScene && currentScene.draw) {
        currentScene.draw(gfx);
    }
}

File: ballattax/BALLATTAX.lua
-- Arcalua version of ballattax

function dump(o)
   if type(o) == 'table' then
      local s = '{ '
      for k,v in pairs(o) do
         if type(k) ~= 'number' then k = '"'..tostring(k)..'"' end
         s = s .. '['..k..'] = ' .. dump(v) .. ', '
      end
      return s .. '}'
   else
      return tostring(o)
   end
end

local currentScene

-- current input state, updated by input events
local inputs = {
    {axes = {0, 0}, buttons = {0, 0, 0, 0, 0, 0, 0, 0}}
}

-- Scene management, dispatch events to game.lua and menu.lua
function switchScene(newScene, args)
    if currentScene and type(currentScene.exit) == "function" then
        currentScene.exit()
    end
    currentScene = newScene
    if currentScene and type(currentScene.enter) == "function" then
        currentScene.enter(args)
    end
end

-- Lifecycle callbacks
function enter(args)
    window.color(0x000000FF) -- Set background color
    switchScene(require(args.scene or "menu"), args)
end

function input(evt,device,id,value,value2)
    if currentScene and currentScene.input then
        currentScene.input(evt,device,id,value,value2)
    end
    -- update inputs state, evt describes a state change:
    if device+1 > #inputs then
        inputs[device+1] = {axes = {0, 0}, buttons = {0, 0, 0, 0, 0, 0, 0, 0}}
    end
    if evt == "axis" then
        inputs[device+1].axes[id+1] = value
    elseif evt == "button" then
        inputs[device+1].buttons[id+1] = value
    end
end

function update(deltaT)
    if currentScene and currentScene.update then
        return currentScene.update(deltaT, inputs)
    end
    return false
end

function draw(gfx)
    if currentScene and currentScene.draw then
        currentScene.draw(gfx)
    end
end
